### PR TITLE
[26724]  Extract Custom Field mixin and subfilters

### PR DESCRIPTION
**Pull request by oliverguenther**
Allows re-use of custom field filters in different contexts, with additional logic per custom field type (e.g., user with me values)

Targeted at 7.4 due to the project filters being added there. We should be able to backport this to 7.3., however we'd need to include the refactoring done for the CFs filter in 7.4.

https://community.openproject.com/wp/26724

### CODE REVIEW

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b8372950076a07112e9c61a00784f92d2b954b46) (deployed commit: 6531bb1fbe528266a2a695689b6733726253856a).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/e0517518307df35a90a321f54ddd50f99340ed8f) (deployed commit: 67b57b8a04ab2e22b90977409bc006f19ec0527c).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ae56d60124bf68965549db110ab670dac8e298b2) (deployed commit: 14678f6b49748269d6f4974c23cbb924947884b1).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/52c6814a9f9c3aa812ef77f606b2bca6c07a9c6b) (deployed commit: 8fa78e4b5ad32096c29d9fdba86f4e6d8c1af040).

File: app/models/queries/filters/serializable.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+require_relative 'base'
+
+module Queries
+  module Filters
+    class Serializable < Base
```
Comment (ulferts) [inline app/models/queries/filters/serializable.rb]: I'd rather have this as a module. As the name already suggests, this is more of a trait rather than an object.

File: app/models/queries/work_packages/filter/me_value_filter_mixin.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      end
+    end
+
+    vals
+  end
+
+  protected
+
+  def me_value
+    'me'.freeze
```
Comment (ulferts) [inline app/models/queries/work_packages/filter/me_value_filter_mixin.rb]: I am aware that this piece of code was just moved but it would make sense to turn this into a constant.

File: spec/features/timelines/view_custom_fields_spec.rb [hunk 3]
```diff
@@ -1,198 +0,0 @@
-#-- copyright
```
Comment (ulferts) [inline spec/features/timelines/view_custom_fields_spec.rb]: Why is this removed within the PR?

**Review by ulferts — COMMENTED** [review]
Looks good, I have just some tiny remarks.

I like that you moved functionality specific for different types of custom fields into separate classes.

File: app/models/queries/work_packages/filter/me_value_filter_mixin.rb (hunk 2 above)
Comment (oliverguenther) [reply app/models/queries/work_packages/filter/me_value_filter_mixin.rb]: Done

File: spec/features/timelines/view_custom_fields_spec.rb (hunk 3 above)
Comment (oliverguenther) [reply spec/features/timelines/view_custom_fields_spec.rb]: By mistake, have re-added it.

File: app/models/queries/filters/serializable.rb (hunk 1 above)
Comment (oliverguenther) [reply app/models/queries/filters/serializable.rb]: Had this as a module, then changed to a base class to make it explicit which filters are serializable through their ancestors. Don't really care though.

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/e22c7c0bffe677bf35b751c718e42ca9f8f50843) (deployed commit: 266bddeeb4dc3fa3d65d1351e432c46ad800a8cb).

**Comment by oliverguenther** [comment]
@ulferts Remaining spec failure is merely flickering
